### PR TITLE
Prevent `XPathEvalError`s when Codelist Mapping XPath identifies non-attribute

### LINKED ISSUE #224: Error in codelist-mapping.xml causing validator problems
There’s a bug with codelists mappings that’s fixed in this PR:
https://github.com/IATI/IATI-Codelists/pull/109

The same bug is repeated here:
https://github.com/IATI/pyIATI/blob/f3c9b54dd3fb6d5e6fc48b3329e6b9a9682d1318/iati/resources/standard/202/codelist-mapping.xml#L60

So `iati.default.codelist_mapping()` loads that… And that’s used by the validator:
https://github.com/IATI/pyIATI/blob/b454bdae51bf706fcb6e46a7c0a0a4b3b27d8e09/iati/validator.py#L287

So when I run `iati.validator.is_valid(dataset, schema)` I hit an error:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".ve/src/pyiati/iati/validator.py", line 693, in is_valid
    return _correct_codelist_values(dataset, schema) and _conforms_with_ruleset(dataset, schema)
  File ".ve/src/pyiati/iati/validator.py", line 500, in _correct_codelist_values
    error_log = _check_codelist_values(dataset, schema)
  File ".ve/src/pyiati/iati/validator.py", line 341, in _check_codelist_values
    error_log.extend(_check_codes(dataset, codelist))
  File ".ve/src/pyiati/iati/validator.py", line 307, in _check_codes
    parents_to_check = dataset.xml_tree.xpath(parent_el_xpath)
  File "src/lxml/etree.pyx", line 2289, in lxml.etree._ElementTree.xpath (src/lxml/etree.c:69770)
  File "src/lxml/xpath.pxi", line 359, in lxml.etree.XPathDocumentEvaluator.__call__ (src/lxml/etree.c:179350)
  File "src/lxml/xpath.pxi", line 227, in lxml.etree._XPathEvaluatorBase._handle_result (src/lxml/etree.c:177714)
lxml.etree.XPathEvalError: Invalid expression
```

Fixing the bug in the codelist mapping appears to fix the problem.

### PR DESCRIPTION
**Pull request by hayfield**
Fixes #224 

It is generally assumed that a Codelist shall be used to check a value in an attribute. This is even stated in the [IATI-Codelists README](https://github.com/IATI/IATI-Codelists#codelist-mapping).

> It's structured as a list of mapping elements, which each have a path element that describes the relevant attribute

Unfortunately, not everything is straightforward in IATIland.

This PR adds support for the current (incorrect) implementation of `2.02` where `//iati-activity/crs-add/channel-code/text()` is an XPath to validate against a Codelist. This may generalise, though should be deemed *highly unstable and a likely candidate for removal* since it is extremely questionable whether this should be permitted.

Something that should be more stable is that a `ValueError` is raised when an unsupported XPath is provided as part of a mapping.

IATI/IATI-Codelists#109 is also integrated.

### CODE REVIEW

File: iati/validator.py [hunk 1]
```diff
@@ -272,6 +272,69 @@ def get_warnings(self):
...
+        dataset (iati.data.Dataset): The Dataset to check Codelist values within.
+        split_xpath (list of str): The sections of the XPath, split by `/`.
+        condition (str): An optional XPath expression to limit the scope of what is extracted.
+
+    Returns:
+        list of tuple: A tuple in the format: `(str, int)` - The `str` is a matching code from within the Dataset; The `int` is the sourceline at which the parent element is located.
+
+    """
+    parent_el_xpath = '/'.join(split_xpath[:-1])
+    last_xpath_section = split_xpath[-1:][0]
```
Comment (andylolz) [inline iati/validator.py]: I think this functionally equivalent to:

```python
last_xpath_section = split_xpath[-1]
```

File: iati/validator.py [hunk 2]
```diff
@@ -290,32 +356,29 @@ def _check_codes(dataset, codelist):
         base_xpath = mapping['xpath']
         condition = mapping['condition']
         split_xpath = base_xpath.split('/')
-        parent_el_xpath = '/'.join(split_xpath[:-1])
-        attr_name = split_xpath[-1:][0][1:]
+        last_xpath_section = split_xpath[-1:][0]
```
Comment (andylolz) [inline iati/validator.py]: Same as https://github.com/IATI/pyIATI/pull/229/files#r149662369 goes for this line.

File: iati/validator.py [hunk 3]
```diff
@@ -272,6 +272,69 @@ def get_warnings(self):
...
+    if condition is None:
+        parent_el_xpath = parent_el_xpath + '[@' + attr_name + ']'
+    else:
+        parent_el_xpath = parent_el_xpath + '[' + condition + ' and @' + attr_name + ']'
+
+    # some nasty string manipulation to make the `//@xml:lang` mapping work
+    while not parent_el_xpath.startswith('//'):
+        parent_el_xpath = '/' + parent_el_xpath
+    if parent_el_xpath.startswith('//['):
+        parent_el_xpath = '//*[' + parent_el_xpath[3:]
```
Comment (andylolz) [inline iati/validator.py]: I guess you’ll hit a merge conflict with #226 here.

File: iati/validator.py [hunk 4]
```diff
@@ -272,6 +272,69 @@ def get_warnings(self):
...
+
+    # include the condition
+    if condition:
+        parent_el_xpath = parent_el_xpath + '[' + condition + ']'
+
+    parents_to_check = dataset.xml_tree.xpath(parent_el_xpath)
+
+    located_codes = list()
+    for parent in parents_to_check:
+        located_codes.append((parent.text, parent.sourceline))
```
Comment (andylolz) [inline iati/validator.py]: I wonder if this should be:

```python
located_codes.append((parent.text.strip(), parent.sourceline))
```

Then the valid test case could be something like:

```xml
<crs-add>
  <channel-code>
    21039
  </channel-code>
</crs-add>
```

File: iati/validator.py (hunk 1 above)
Comment (andylolz) [reply iati/validator.py]: I.e. the code that’s there makes a list of all items from the last one onwards, then takes the first item of that list.

File: iati/validator.py (hunk 4 above)
Comment (hayfield) [reply iati/validator.py]: Matching of Codelists in attributes is exact, with whitespace being deemed incorrect. As such, stripping whitespace for element text doesn't seem like a correct interpretation. Maybe a warning?

File: iati/validator.py (hunk 1 above)
Comment (hayfield) [reply iati/validator.py]: So it is!

Comment (hayfield) [reply iati/validator.py]: d12d03fc196406326b43fd59f1c1fc366d558b4b

File: iati/validator.py (hunk 2 above)
Comment (hayfield) [reply iati/validator.py]: d12d03fc196406326b43fd59f1c1fc366d558b4b

File: iati/validator.py (hunk 4 above)
Comment (andylolz) [reply iati/validator.py]: So I agree for attributes… I’d be tempted to be more forgiving for element text. As a quite hand-wavy justification, IATI-Rulesets allows for whitespace around an `iati-identifier`:

 * https://github.com/IATI/IATI-Rulesets/blob/c33d11885e1084bb99d897c9a5dc3bfcea91c0fb/iatirulesets/__init__.py#L63
 * https://github.com/IATI/IATI-Rulesets/blob/c33d11885e1084bb99d897c9a5dc3bfcea91c0fb/rulesets/standard.json#L19-L20

…so the following passes:

```xml
<iati-identifier>
  some-valid-identifier
</iati-identifier>
```

Comment (hayfield) [reply iati/validator.py]: > As a quite hand-wavy justification, IATI-Rulesets allows for whitespace around an iati-identifier

In that same implementation, [startswith](https://github.com/IATI/IATI-Rulesets/blob/c33d11885e1084bb99d897c9a5dc3bfcea91c0fb/iatirulesets/__init__.py#L74) [doesn't permit whitespace](https://github.com/IATI/IATI-Rulesets/blob/c33d11885e1084bb99d897c9a5dc3bfcea91c0fb/iatirulesets/__init__.py#L11) leading whitespace.

---

The source to go back to is the [XPath spec](https://www.w3.org/TR/1999/REC-xpath-19991116/#node-tests) which shows that `text()` returns the entire [text node](https://www.w3.org/TR/1999/REC-xpath-19991116/#section-Text-Nodes). The XPath in the mapping file says that the comparison is against this value (rather than a trimmed or stripped version). As such, checking against a modified text node is not correct.

Further, the [XML Spec](https://www.w3.org/TR/xml/#sec-white-space) shows that white space is significant and must be treated as part of a value - treating it as not-present is incorrect.

Comment (andylolz) [reply iati/validator.py]: > In that same implementation, startswith doesn't permit whitespace leading whitespace.

Well, that’s not used in the standard ruleset… But yes – I take your point and I do agree!

It looks like slovakaid are the only publisher currently using the element anyway, and [they get it right](https://iatiregistry.org/dataset/slovakaid-69_1_ac). Also, very few publishers seem to make the mistake of whitespace around `iati-identifier`s.

**Comment by andylolz** [comment]
> It is generally assumed that a Codelist shall be used to check a value in an attribute. This is even stated in the IATI-Codelists README.

Rather than updating the schema, couldn’t you just update the IATI-Codelists README?

Put another way: It seems we’re treating text nodes exactly like attributes… So can you explain why it’s a problem to use text nodes for codelist items?

**Comment by hayfield** [comment]
> So can you explain why it’s a problem to use text nodes for codelist items?

It's an unnecessary special case that complicates data use.

The purpose of a Codelist is to restrict the values to a list that can be used between different systems. A Code has multiple components - a `code`, `name`, `description` and more. Some of these, such as `name` and `description` can be long portions of text. The `code`, however, is a short language-independent value - generally a numeric value, though potentially with a letter or two as a prefix.

The only notable (for this use case) difference between the permitted content of elements and attributes is the ability to maintain newlines. There are a grand total of zero cases within the IATI Standard where it could possibly be necessary for a Code's `code` component to contain a newline. Even with namespaces, there is zero need if you're using Codelists correctly.

The one potentially arguable case is `code`s in external vocabularies where an external organisation maintains the code. All IATI elements with a `@vocabulary-uri` attribute require that the value from this fits within a `@code` *attribute*. Additionally, no relevant Replicated Codelist includes newlines in any `code` values. As such, this doesn't appear a relevant argument.

Of all these cases within the Standard itself, there are zero reasons to use an element's text node over an attribute. All it does is complicate data use by having a single case that breaks all other conventions (IATI has quite a few of these).

---

External to the Standard itself, the situation where this might possibly be relevant is that you wish to validate `//narrative/text()` against fixed vocabularies (checking against some internal system or something). A custom Codelist would be a manner that could be undertaken since there's not really anything better.

---

Basically, The Standard has zero need to use an element's text node over an attribute. It could, however, be beneficial in a wider content-validation sense, though this doesn't appear to be permitted within the current specification of components and should be discussed separately.

**Comment by andylolz** [comment]
> Of all these cases within the Standard itself, there are zero reasons to use an element's text node over an attribute.

Yeah, agreed – that makes sense to me.

I take from this that while there’s nothing technically wrong here, it’s needlessly inconsistent, which is bad.

That is: if a standard used text nodes for codelist items instead of attributes – but did it consistently – that would be a reasonable design choice. But mixing the two is bad. Is that fair?

**Comment by andylolz** [comment]
> Of all these cases within the Standard itself, there are zero reasons to use an element's text node over an attribute.

Yeah, agreed – that makes sense to me.

I take from this that while there’s nothing technically wrong here, it’s needlessly inconsistent, which is bad.

That is: if a standard were to use text nodes for codelist items instead of attributes – but did it consistently – that would be a reasonable design choice. But mixing the two is bad. Is that fair?

File: iati/validator.py [hunk 5]
```diff
@@ -290,32 +356,29 @@ def _check_codes(dataset, codelist):
...
-                    error = ValidationError('err-code-not-on-codelist', locals())
+                if last_xpath_section.startswith('@'):
+                    attr_name = last_xpath_section[1:]  # used via `locals()` # pylint: disable=unused-variable
+                    if codelist.complete:
+                        error = ValidationError('err-code-not-on-codelist', locals())
+                    else:
+                        error = ValidationError('warn-code-not-on-codelist', locals())
                 else:
-                    error = ValidationError('warn-code-not-on-codelist', locals())
+                    el_name = split_xpath[-2:-1][0]  # used via `locals()` # pylint: disable=unused-variable
```
Comment (andylolz) [inline iati/validator.py]: It takes me a while to get my head around this notation every time! I’d write this as:

```python
el_name = split_xpath[-2]
```

Comment (andylolz) [reply iati/validator.py]: More radical suggestion: I actually think this whole bit would be more legible if you scrapped `split_xpath`, used `rsplit('/', 1)` all over the place, and passed strings around instead of lists of strings. So like:

```python
parent_el_xpath, last_xpath_section = base_xpath.rsplit('/', 1)
```

But do feel free to ignore this suggestion, because this pull request looks good as it is!

**Comment by hayfield** [comment]
> That is: if a standard were to use text nodes for codelist items instead of attributes – but did it consistently – that would be a reasonable design choice. But mixing the two is bad. Is that fair?

Seems like a good assessment of the situation, yep.

File: iati/validator.py (hunk 5 above)
Comment (hayfield) [reply iati/validator.py]: https://github.com/IATI/pyIATI/pull/229/commits/ec3f996030a7d4bfff6989b157ffc32773e9ede3

(I clearly haven't properly got my head around negative array indexing >.< - super useful feature that few languages I've used much have)

Comment (hayfield) [reply iati/validator.py]: With the `rsplit` version, it would require `_extract_codes_from_attrib()` and `_extract_codes_from_element_text()` to have different argument lists.

If these functions were public, that wouldn't be a great thing (lack of consistency). Since they're private, this isn't ideal, though is less of a problem.

It would, however, make the code a bit more DRY since `last_xpath_section` and `parent_el_xpath` would each only need calculating once - in `_check_codes()`. Hum.

Comment (andylolz) [reply iati/validator.py]: ¯\_(ツ)_/¯

Comment (hayfield) [reply iati/validator.py]: 0ec07bb265bac5b7ae37b08940c8a58d406c8cab

Comment (andylolz) [reply iati/validator.py]: Different args for those functions seems intuitively correct: one needs to know am attribute reference; the other doesn't.

I will stop peddling my reckons on this now, because I'm unsure so I trust whatever you decide!

File: iati/validator.py [hunk 6]
```diff
@@ -282,43 +343,37 @@ def _check_codes(dataset, codelist):
...
-        # provide a secondary cludge to deal with the 'xml' namespace
-        if attr_name == 'xml:lang':
-            attr_name = '{http://www.w3.org/XML/1998/namespace}lang'
+        parent_el_xpath, last_xpath_section = mapping['xpath'].rsplit('/', 1)
 
-        parents_to_check = dataset.xml_tree.xpath(parent_el_xpath)
-
-        for parent in parents_to_check:
-            code = parent.attrib[attr_name]
+        if last_xpath_section.startswith('@'):
```
Comment (allthatilk) [inline iati/validator.py]: This should be a function called `_extract_codes`

File: iati/validator.py [hunk 7]
```diff
@@ -272,6 +272,67 @@ def get_warnings(self):
...
+    parents_to_check = dataset.xml_tree.xpath(parent_el_xpath)
+
+    located_codes = list()
+    for parent in parents_to_check:
+        located_codes.append((parent.text, parent.sourceline))
+
+    return located_codes
+
+
 def _check_codes(dataset, codelist):
```
Comment (allthatilk) [inline iati/validator.py]: This badly needs refactoring. Function neither DRY nor SRP.

File: iati/validator.py [hunk 8]
```diff
@@ -282,43 +343,37 @@ def _check_codes(dataset, codelist):
...
+            raise ValueError('mapping path does not locate attribute value or element text')
 
+        for (code, line_number) in located_codes:  # `line_number` used via `locals()` # pylint: disable=unused-variable
             if code not in codelist.codes:
-                line_number = parent.sourceline  # used via `locals()` # pylint: disable=unused-variable
-
-                if codelist.complete:
-                    error = ValidationError('err-code-not-on-codelist', locals())
+                if last_xpath_section.startswith('@'):
+                    if codelist.complete:
```
Comment (allthatilk) [inline iati/validator.py]: extract setting of error name to a variable

Comment (hayfield) [reply iati/validator.py]: c2df432c4a955e99eeb8c83465bd6c47fa041f8e
